src/types/poly_func.rs

```diff
@@ -0,0 +1,613 @@
+//! Polymorphic Function Types
+
+use crate::{
+    extension::{ExtensionRegistry, SignatureError},
+    types::type_param::check_type_arg,
+};
+use itertools::Itertools;
+
+use super::type_param::{check_type_args, TypeArg, TypeParam};
+use super::{FunctionType, Substitution};
+
+/// A polymorphic function type, e.g. of a [Graph], or perhaps an [OpDef].
+/// (Nodes/operations in the Hugr are not polymorphic.)
+///
+/// [Graph]: crate::values::PrimValue::Function
+/// [OpDef]: crate::extension::OpDef
+#[derive(
+    Clone, PartialEq, Debug, Eq, derive_more::Display, serde::Serialize, serde::Deserialize,
+)]
+#[display(
+    fmt = "forall {}. {}",
+    "params.iter().map(ToString::to_string).join(\" \")",
+    "body"
+)]
+pub struct PolyFuncType {
+    /// The declared type parameters, i.e., these must be instantiated with
+    /// the same number of [TypeArg]s before the function can be called. Note that within
+    /// the [Self::body], variable (DeBruijn) index 0 is element 0 of this array, i.e. the
```

Comment:
This will be very familiar to BRAT developers as the Bwd list...

Reply:
I guess the other way would be to treat all variable indices as negative offsets from the end, but that seems pretty nasty

Reply:
I agree, using the variable as an index into the list is probably more natural. I feel the reverse indexing would be more surprising

Reply:
Yes, so branch `new/poly_types_varidx` manages to hide all that away behind the VarIdx struct, and it is largely restricted to the `types` module - HOWEVER anyone building PolyFuncTypes will have to work with indices directly, so unless we offer Yaml parsing as the only interface, the L-R / R-L indexing is exposed to endusers, and I am leaning towards the current scheme.

I'll put the VarIdx changes *without changing indexing* up as a separate "refactor" PR so we can consider them later if we want, but it seems ok to do that as a second step.